text/3874-build-std-always.md

```diff
@@ -0,0 +1,1543 @@
+    archived and 131 MiB extracted.
+- Vendored dependencies can't be updated with the latest security fixes
+  - This is no different than the pre-built standard library
+
+How this affects `crates.io`/`rustup` bandwidth usage or user time spent
+downloading these crates is unclear and depends on user patterns. If not
+vendored, Cargo will "lazily" download them the first time `build-std` is used
+but this may happen multiple times if they are cleaned from its cache without
+upgrading the toolchain version.
+
+See
+[*Why use the lockfile of the `rust-src` component?*][rationale-lockfile]
+
+↩ [*Vendored `rust-src`*][vendored-rust-src]
+
+## Why not check if `rust-src` has been modified?
+[rationale-src-modifications]: #why-not-check-if-rust-src-has-been-modified
+
+This is in line with other immutable dependency sources (like registry or git).
+It is also likely that any protections implemented to check that the sources in
+`rust-src` have not been modified could be trivially bypassed.
+
+Any crate that depends on `rust-src` having been modified would not be usable
+when published to crates.io as the required modifications will obviously not be
+included.
+
+↩ [*Vendored `rust-src`*][vendored-rust-src]
+
+## Why allow building from the sysroot with implied `RUSTC_BOOTSTRAP`?
+[rationale-implied-bootstrap]: #why-allow-building-from-the-sysroot-with-implied-rustc_bootstrap
+
+Cargo needs to be able to build the standard library crates, which inherently
+require unstable features. It could set `RUSTC_BOOTSTRAP` internally to do this
+with a stable toolchain, but this is a bypass mechanism that the project do not
+want to encourage use of, and as this is a shared requirement with other build
+systems that wish to build an unmodified standard library and want to work on
+stable toolchains, it is worth establishing a narrow general mechanism.
+
+For example, Rust's project goal to enable Rust for Linux to build using only a
+stable toolchain would require that it be possible to build `core` without
+nightly.
+
+It is not sufficient for rustc to special-case the `core`, `alloc` and `std`
+crate names as, when being built as part of the standard library, dependencies
+of the standard library also use unstable features and it is not practical to
+special-case all of these crates.
+
+↩ [*Building the standard library on a stable toolchain*][building-the-standard-library-on-a-stable-toolchain]
+
+## Why invert the `mem` feature?
+[rationale-no-mem]: #why-invert-the-mem-feature
+
+While "negative" features are typically discouraged due to how features unify
+(e.g. `std` features are preferred to `no_std`): the `mem` feature's current
+behaviour is the opposite of what is optimal.
+
+Ideally, a crate should be able to provide alternate memory symbols and disable
+`compiler_builtins`' symbols for the entire crate graph by enabling a feature
+(e.g. `std`/`libc` could do this) - this is what an `external-mem` feature
+enables.
```

Comment:
Until external-mem automatically gets enabled for all targets with a libc implementation, this inversion of the mem feature would cause a silent performance hit for those targets, while previously you would get a loud failure, which IMO is preferable.